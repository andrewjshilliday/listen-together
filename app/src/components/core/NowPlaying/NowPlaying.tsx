import React from 'react';
import { Link } from 'react-router-dom';
import { CircularProgress, Slider } from '@material-ui/core';
import moment from 'moment';
import { useMusicKit } from '../../providers';
import { MusicKitService } from '../../../services';
import styled from 'styled-components';

type RepeatMode = 0 | 1 | 2;

const NowPlaying: React.FC = (props: any) => {
  const musicKitProvider = useMusicKit();
  
  const handleVolumeChange = (event: any, newValue: number | number[]) => {
    musicKitProvider.actions.setVolume(newValue as number);
  };

  return (
    <PlayerBar playing={musicKitProvider.nowPlayingItem != null}>
      {musicKitProvider.nowPlayingItem ?
        <>
          <NowPlayingInfoContainer>
            <Link to={`/album/${musicKitProvider.nowPlayingItem.assets[0].metadata.playlistId}`}>
              <NowPlayingArtwork src={MusicKitService.FormatArtwork(musicKitProvider.nowPlayingItem.attributes.artwork, 70)} className="rounded" alt={musicKitProvider.nowPlayingItem.attributes.name} />
            </Link>
            <NowPlaingText>
              <span className="text-truncate">{musicKitProvider.nowPlayingItem.attributes.name}</span>
              <Link to={`/album/${musicKitProvider.nowPlayingItem.assets[0].metadata.playlistId}`}><span className="text-truncate">{musicKitProvider.nowPlayingItem.attributes.albumName}</span></Link>
              <Link to={`/artist/${musicKitProvider.nowPlayingItem.assets[0].metadata.artistId}`}><span className="text-truncate">{musicKitProvider.nowPlayingItem.attributes.artistName}</span></Link>
            </NowPlaingText>
          </NowPlayingInfoContainer>
          <MediaControlsContainer>
            <MediaControlsButtonContainer>
              {/* {musicKitProvider.musicKit.player.repeatMode === 0 && <RepeatIcon repeatMode={musicKitProvider.musicKit.player.repeatMode}><i className="fas fa-redo" onClick={() => {return}}></i></RepeatIcon>}
              {musicKitProvider.musicKit.player.repeatMode === 1 && <RepeatIcon repeatMode={musicKitProvider.musicKit.player.repeatMode}><i className="fas fa-redo" onClick={() => {return}}></i></RepeatIcon>}
              {musicKitProvider.musicKit.player.repeatMode === 2 && <RepeatIcon repeatMode={musicKitProvider.musicKit.player.repeatMode}><i className="fas fa-redo" onClick={() => {return}}></i></RepeatIcon>} */}
              <PreviousIcon><i className="fas fa-backward" onClick={() => musicKitProvider.actions.previous()}></i></PreviousIcon>
              <PlayIcon>
                {!musicKitProvider.isPlaying && !musicKitProvider.playbackLoading && <i className="fas fa-play" onClick={() => musicKitProvider.actions.play()}></i>}
                {musicKitProvider.isPlaying && !musicKitProvider.playbackLoading && <i className="fas fa-pause" onClick={() => musicKitProvider.actions.pause()}></i>}
                {musicKitProvider.playbackLoading && <PlaybackLoadingSpinner onClick={() => musicKitProvider.actions.stop()}></PlaybackLoadingSpinner>}
              </PlayIcon>
              <NextIcon><i className="fas fa-forward" onClick={() => musicKitProvider.actions.next()}></i></NextIcon>
              {/* <ShuffleIcon shuffleMode={musicKitProvider.musicKit.player.shuffleMode}><i className="fas fa-random" onClick={() => {return}}></i></ShuffleIcon> */}
            </MediaControlsButtonContainer>
            <MediaProgressBar>
              <CurrentPlaybackTime>{moment.utc(musicKitProvider.currentPlaybackTime*1000).format('mm:ss')}</CurrentPlaybackTime>
              <Slider value={musicKitProvider.currentPlaybackTime} onChangeCommitted={(e, v) => musicKitProvider.actions.seek(v)} max={musicKitProvider.musicKit.player.currentPlaybackDuration}></Slider>
              <CurrentPlaybackDuration>-{moment.utc(musicKitProvider.currentPlaybackTimeRemaining*1000).format('m:ss')}</CurrentPlaybackDuration>
            </MediaProgressBar>
          </MediaControlsContainer>
          <ExtraControlsContainer>
            <VolumeSlider>
              <i className="fas fa-volume-down"></i>
              <Slider max={1} min={0} step={0.01} value={musicKitProvider.musicKit.player.volume} onChange={handleVolumeChange}></Slider>
              <i className="fas fa-volume-up"></i>
            </VolumeSlider>
          </ExtraControlsContainer>
        </>
        :
        <h1>ListenTogether</h1>
      }
    </PlayerBar>
  );
}

export default NowPlaying;


const PlayerBar = styled.div<{playing: boolean}>`
  position: absolute;
  display: flex;
  background-color: var(--background-darker);
  color: var(--on-background);
  bottom: 0;
  border-top: 1px solid var(--on-background);
  width: 100%;
  height: 100px;

  ${props => !props.playing && `
    justify-content: center;
    align-items: center;
  `}
`;
const NowPlayingInfoContainer = styled.div`
  display: flex;
  align-items: center;
  width: 25%;
  min-width: 270px;
  padding: 0 15px;

  .now-playing-text {
  }
`;
const NowPlayingArtwork = styled.img`
  width: 70px;
  height: 70px;
`;
const NowPlaingText = styled.div`
  margin-left: 10px;
  overflow: hidden;
  span { display: block; }
`;
const MediaControlsContainer = styled.div`
  display: block;
  width: 50%;
`;
const MediaControlsButtonContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  padding-top: 11px;

  i {
    cursor: pointer;
  }
`;
const PlayIcon = styled.div`
  padding: 0 20px;
  i {
    color: var(--on-background);
    font-size: 40pt;
    &:hover { color: var(--primary); }
  }
`;
const PlaybackLoadingSpinner = styled(CircularProgress)`
  margin: 7px 3px 0 3px;
`;
const PreviousIcon = styled.div`
  i {
    color: var(--on-background);
    font-size: 30pt;
    &:hover { color: var(--primary); }
  }
`;
const NextIcon = styled.div`
  i {
    color: var(--on-background);
    font-size: 30pt;
    &:hover { color: var(--primary); }
  }
`;
const RepeatIcon = styled.div<{repeatMode: number}>`
  padding: 0 15px;
  i {
    color: var(--on-background);
    font-size: 15pt;
  }
`;
const ShuffleIcon = styled.div<{shuffleMode: number}>`
  padding: 0 15px;
  i {
    color: var(--on-background);
    font-size: 15pt;
  }
`;
const MediaProgressBar = styled.div`
  display: flex;
  padding: 0 20px;
`;
const CurrentPlaybackTime = styled.span`
  position: relative;
  top: 1px;
  color: var(--on-background);
  width: 60px;
  font-size: 11pt;
`;
const CurrentPlaybackDuration = styled(CurrentPlaybackTime)`
  text-align: end;
`;
const ExtraControlsContainer = styled.div`
  display: flex;
  width: 25%;
  min-width: 270px;
  justify-content: flex-end;
`;
const VolumeSlider = styled.div`
  display: flex;
  width: 70%;
  align-items: center;
  i { padding: 0 15px; }
`;
