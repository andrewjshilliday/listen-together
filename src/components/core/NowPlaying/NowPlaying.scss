.player-bar {
  position: absolute;
  display: flex;
  background-color: var(--background-darker);
  color: var(--on-background);
  bottom: 0;
  border-top: 1px solid black;
  width: 100%;
  height: 100px;

  &.not-playing {
    justify-content: center;
    align-items: center;
  }
}

.now-playing-info {
  display: flex;
  align-items: center;
  width: 25%;
  min-width: 270px;
  padding: 0 15px;

  .artwork {
    width: 70px;
    height: 70px;
  }

  .now-playing-text {
    margin-left: 10px;
    overflow: hidden;
    span { display: block; }
  }
}

.media-controls {
  display: block;
  width: 50%;

  .media-control-buttons {
    display: flex;
    justify-content: center;
    align-items: center;
    padding-top: 11px;

    .play {
      padding: 0 20px;
      i {
        color: var(--on-background);
        font-size: 40pt;
        &:hover { color: var(--primary); }
      }
    }
    .previous, .next {
      i {
        color: var(--on-background);
        font-size: 30pt;
        &:hover { color: var(--primary); }
      }
    }
    .repeat, .shuffle {
      padding: 0 15px;
      i {
        color: var(--on-background);
        font-size: 15pt;
      }
    }

    i {
      cursor: pointer;
    }

    .spinner {
      margin: 7px 3px 0 3px;
    }
  }

  .media-progress-bar {
    display: flex;
    padding: 0 20px;

    .current-playback-time, .current-playback-duration {
      position: relative;
      top: 1px;
      color: black;
      width: 60px;
      font-size: 11pt;
    }

    .current-playback-duration {
      text-align: end;
    }
  }
}

.extra-controls {
  display: flex;
  width: 25%;
  min-width: 270px;
  justify-content: flex-end;

  .volume-slider {
    display: flex;
    width: 70%;
    align-items: center;
    i { padding: 0 15px; }
  }
}
