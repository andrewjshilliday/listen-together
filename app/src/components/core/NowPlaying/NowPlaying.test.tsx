import React from 'react';
import { StaticRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import NowPlaying from './NowPlaying';

jest.mock('../../../services/MusicKit');
jest.mock('../../providers/MusicKitProvider', () => require('../../../__mocks__/providers/MusicKitProvider.mock').default);

it('renders correctly', () => {
  const tree = renderer.create(
    <StaticRouter location=".">
      <NowPlaying />
    </StaticRouter>).toJSON();
  expect(tree).toMatchSnapshot();
});