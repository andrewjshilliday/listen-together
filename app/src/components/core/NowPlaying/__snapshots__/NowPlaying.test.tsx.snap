// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-AxjAm jmalSx"
>
  <div
    className="sc-AxirZ gxnMoX"
  >
    <a
      href="/album/undefined"
      onClick={[Function]}
    >
      <img
        alt="This Land"
        className="sc-AxiKw hLDHux rounded"
      />
    </a>
    <div
      className="sc-AxhCb iXIybJ"
    >
      <span
        className="text-truncate"
      >
        This Land
      </span>
      <a
        href="/album/undefined"
        onClick={[Function]}
      >
        <span
          className="text-truncate"
        >
          This Land
        </span>
      </a>
      <a
        href="/artist/undefined"
        onClick={[Function]}
      >
        <span
          className="text-truncate"
        >
          Gary Clark Jr.
        </span>
      </a>
    </div>
  </div>
  <div
    className="sc-AxhUy dyydcY"
  >
    <div
      className="sc-AxgMl hbtauS"
    >
      <div
        className="sc-AxmLO hDKKzl"
      >
        <i
          className="fas fa-backward"
          onClick={[Function]}
        />
      </div>
      <div
        className="sc-AxheI exHRBg"
      >
        <i
          className="fas fa-pause"
          onClick={[Function]}
        />
      </div>
      <div
        className="sc-fzozJi bwYmJg"
      >
        <i
          className="fas fa-forward"
          onClick={[Function]}
        />
      </div>
    </div>
    <div
      className="sc-fzplWN dAveBb"
    >
      <span
        className="sc-fznyAO bqCRiV"
      >
        00:00
      </span>
      <span
        className="MuiSlider-root MuiSlider-colorPrimary"
        onMouseDown={[Function]}
      >
        <span
          className="MuiSlider-rail"
        />
        <span
          className="MuiSlider-track"
          style={
            Object {
              "left": "0%",
              "width": "0%",
            }
          }
        />
        <input
          type="hidden"
          value="0"
        />
        <span
          aria-orientation="horizontal"
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={0}
          className="MuiSlider-thumb MuiSlider-thumbColorPrimary"
          data-index={0}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          role="slider"
          style={
            Object {
              "left": "0%",
            }
          }
          tabIndex={0}
        />
      </span>
      <span
        className="sc-fznyAO sc-fznKkj hMvlIi"
      >
        -
        1:00
      </span>
    </div>
  </div>
  <div
    className="sc-fznZeY sJXdW"
  >
    <div
      className="sc-fzokOt geWHSN"
    >
      <i
        className="fas fa-volume-down"
      />
      <span
        className="MuiSlider-root MuiSlider-colorPrimary"
        onMouseDown={[Function]}
      >
        <span
          className="MuiSlider-rail"
        />
        <span
          className="MuiSlider-track"
          style={
            Object {
              "left": "0%",
              "width": "100%",
            }
          }
        />
        <input
          type="hidden"
          value="1"
        />
        <span
          aria-orientation="horizontal"
          aria-valuemax={1}
          aria-valuemin={0}
          aria-valuenow={1}
          className="MuiSlider-thumb MuiSlider-thumbColorPrimary"
          data-index={0}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          role="slider"
          style={
            Object {
              "left": "100%",
            }
          }
          tabIndex={0}
        />
      </span>
      <i
        className="fas fa-volume-up"
      />
    </div>
  </div>
</div>
`;
